### PR TITLE
libobs: Remove excessive null checks

### PR DESCRIPTION
**Pull request by notr1ch**
### Description
Various libobs functions check for a null obs_core. Several of the checks dereference the pointer before checking it which would result in a crash if null were ever passed, and the null checks are not done consistently across all functions. Assuming obs_startup and obs_shutdown are called properly, there should never be a null obs_core in libobs. This PR removes the null checks.

This requires a slight re-ordering of the shutdown code, as modules which call into libobs during unload might have been accessing one of the functions protected with a null check, and obs_core is actually null at that point. The modules now unload while obs_core is still valid.

### Motivation and Context
Performance and code cleanup.

### How Has This Been Tested?
Ran OBS through my usual testing, no issues encountered. Small risk that some 3rd party plugins may be calling libobs incorrectly and would now crash instead of the call being a no-op.

### Types of changes

- Code cleanup (non-breaking change which makes code smaller or more readable)
- Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

File: libobs/obs.c [hunk 1]
```diff
@@ -1027,6 +1023,14 @@ void obs_shutdown(void)
...
 	stop_hotkeys();
 
+	module = obs->first_module;
+	while (module) {
+		struct obs_module *next = module->next;
+		free_module(module);
+		module = next;
+	}
+	obs->first_module = NULL;
+
```
Comment (palana) [inline libobs/obs.c:1033]: This seems to go beyond "Remove excessive null checks", should probably be in another commit (e.g. the behavior change seems to be too big to be buried in the last paragraph of a commit message with a different title)

Comment (notr1ch) [reply libobs/obs.c:1033]: I've split this into two separate commits now.

**Comment by notr1ch** [comment]
To try and ease the concerns with the plugin unload changes, I did some additional testing by catching libobs calls during unload. Only a single core plugin, obs-scripting, calls into libobs during unload, specifically `obs_remove_tick_callback` which works fine with the changes.

I also tested some popular 3rd party plugins: StreamFX calls a lot of libobs functions in destructors to clean up resources. obs-websocket calls functions such as `obs_get_signal_handler` to clean up signals. obs-ndi and obs-virtualcam do not perform any calls. No issues were encountered with StreamFX or obs-websocket calling with an active obs_core.

**Comment by WizardCM** [comment]
Ran into a crash: when you open a second instance while OBS is currently running, pressing "Cancel" in the "OBS is already running" dialog causes a crash.

https://github.com/obsproject/obs-studio/blob/bb890a674e8289f7be2825fbeea13c7d2f9623af/libobs/obs.c#L997
> `obs` was nullptr

**Comment by notr1ch** [comment]
This PR was intended to expose such bugs and seems to be doing its job :). In this particular code path, cancelling the launch skips the call to obs_startup, but the UI still calls obs_shutdown. Will look into fixing that.
